### ISSUE TITLE
Add Reinforcement Learning (RL) Integration for MELD Simulations

### ISSUE THREAD

**Issue by ctava-msft** [ctava-msft/d3-meld#1]
Summary
Introduce a reinforcement learning (RL) control loop into the MELD simulation workflow by creating a Gymnasium-compatible environment and integrating an RL agent (starting with PPO from Stable Baselines3). This will allow adaptive steering of MELD simulations to improve sampling efficiency.

Background
Current MELD runs use fixed protocols for replica exchange and biasing. Adding RL provides a way to learn adaptive strategies based on simulation feedback. The system runs on 4×A100 GPUs:

GPU 0: Leader process
GPUs 1–3: 29 worker replicas

The RL agent will run on GPU 3 without disrupting existing parallelism.

Proposed Changes

Create a Gymnasium Environment Wrapper

Implement MeldEnv class with:

reset(): Initialize MELD simulation.
step(action): Apply RL action, run simulation for a block, return observation and reward.
Define observation_space and action_space.

Patch MELD for RL Hooks

Add Python API hooks to:

Pause simulation after a block.
Modify simulation parameters dynamically.

Integrate RL Framework

Start with Stable Baselines3 (PPO).
Configure GPU allocation: RL agent on GPU 3, MELD replicas on GPUs 1–2.

Add Training Script

train_rl_agent.py to run PPO training with MeldEnv.

Documentation

Update README with setup instructions and example workflow.

Acceptance Criteria

 MeldEnv implements Gymnasium API and runs a minimal MELD simulation loop.
 MELD accepts runtime parameter changes without restart.
 PPO agent completes at least one training episode without errors.
 RL agent runs on GPU 3 while MELD replicas run on GPUs 1–2.
 Training logs metrics to TensorBoard.
 README includes RL integration instructions.

import gymnasium as gym
from gymnasium import spaces
import numpy as np

class MeldEnv(gym.Env):
    """
    Custom Gymnasium environment for MELD simulations.
    """
    metadata = {"render_modes": ["human"]}

    def __init__(self, config):
        super(MeldEnv, self).__init__()
        self.config = config
        # Example: observation = [RMSD, energy]
        self.observation_space = spaces.Box(low=-np.inf, high=np.inf, shape=(2,), dtype=np.float32)
        # Example: action = adjust bias strength (continuous)
        self.action_space = spaces.Box(low=0.0, high=1.0, shape=(1,), dtype=np.float32)
        self.simulation = None  # Placeholder for MELD simulation object

    def reset(self, seed=None, options=None):
        super().reset(seed=seed)
        self.simulation = self._init_simulation()
        obs = self._get_observation()
        return obs, {}

    def step(self, action):
        self._apply_action(action)
        self._run_simulation_block()
        obs = self._get_observation()
        reward = self._compute_reward(obs)
        done = self._check_done(obs)
        return obs, reward, done, False, {}

    def render(self):
        print(f"Current state: {self._get_observation()}")

    # Internal helpers
    def _init_simulation(self):
        pass  # Initialize MELD system

    def _apply_action(self, action):
        pass  # Modify MELD parameters dynamically

    def _run_simulation_block(self):
        pass  # Advance simulation by BLOCK_SIZE steps

    def _get_observation(self):
        return np.array([0.0, 0.0], dtype=np.float32)

    def _compute_reward(self, obs):
        return -obs[0]  # Example: negative RMSD

    def _check_done(self, obs):
        return False

import torch
from stable_baselines3 import PPO
from stable_baselines3.common.callbacks import CheckpointCallback
from meld_env import MeldEnv

if __name__ == "__main__":
    # Config for MELD simulation
    config = {
        "n_replicas": 30,
        "block_size": 50,
        "timesteps": 25000,
        "minimize_steps": 20000,
        "gpu_id": 3  # RL agent GPU
    }

    # Create environment
    env = MeldEnv(config)

    # PPO model (policy on GPU 3)
    device = torch.device("cuda:0")  # Ensure CUDA_VISIBLE_DEVICES maps GPU 3
    model = PPO("MlpPolicy", env, verbose=1, device=device, tensorboard_log="./ppo_meld_tensorboard/")

    # Callbacks for checkpoints
    checkpoint_callback = CheckpointCallback(save_freq=10000, save_path="./checkpoints/", name_prefix="ppo_meld")

    # Train
    model.learn(total_timesteps=100000, callback=checkpoint_callback)

    # Save final model
    model.save("ppo_meld_final")
